source/extensions/matching/network/common/inputs.h

```diff
@@ -256,30 +256,47 @@ class TransportProtocolInputFactory
 
 DECLARE_FACTORY(TransportProtocolInputFactory);
 
-class FilterStateInput : public Matcher::DataInput<MatchingData> {
+template <class MatchingDataType>
+class FilterStateInput : public Matcher::DataInput<MatchingDataType> {
 public:
-  FilterStateInput(
-      const envoy::extensions::matching::common_inputs::network::v3::FilterStateInput& input_config)
-      : filter_state_key_(input_config.key()) {}
+  FilterStateInput(const std::string& filter_state_key) : filter_state_key_(filter_state_key) {}
 
-  Matcher::DataInputGetResult get(const MatchingData& data) const override;
+  Matcher::DataInputGetResult get(const MatchingDataType& data) const override {
+    const auto* filter_state_object =
+        data.filterState().template getDataReadOnly<StreamInfo::FilterState::Object>(
+            filter_state_key_);
+
+    if (filter_state_object != nullptr) {
+      auto str = filter_state_object->serializeAsString();
+      if (str.has_value()) {
+        return {Matcher::DataInputGetResult::DataAvailability::AllDataAvailable, str.value()};
+      } else {
+        return {Matcher::DataInputGetResult::DataAvailability::AllDataAvailable, absl::monostate()};
+      }
+    }
+
+    return {Matcher::DataInputGetResult::DataAvailability::AllDataAvailable, absl::monostate()};
+  }
 
 private:
   const std::string filter_state_key_;
 };
 
-class FilterStateInputFactory : public Matcher::DataInputFactory<MatchingData> {
+template <class MatchingDataType>
+class FilterStateInputBaseFactory : public Matcher::DataInputFactory<MatchingDataType> {
 public:
   std::string name() const override { return "envoy.matching.inputs.filter_state"; }
 
-  Matcher::DataInputFactoryCb<MatchingData> createDataInputFactoryCb(
-      const Protobuf::Message& message,
-      ProtobufMessage::ValidationVisitor& message_validation_visitor) override {
-    const auto& proto_config = MessageUtil::downcastAndValidate<
+  Matcher::DataInputFactoryCb<MatchingDataType>
+  createDataInputFactoryCb(const Protobuf::Message& message,
+                           ProtobufMessage::ValidationVisitor& validation_visitor) override {
+    const auto& typed_config = MessageUtil::downcastAndValidate<
         const envoy::extensions::matching::common_inputs::network::v3::FilterStateInput&>(
-        message, message_validation_visitor);
+        message, validation_visitor);
 
-    return [proto_config]() { return std::make_unique<FilterStateInput>(proto_config); };
+    return [filter_state_key = typed_config.key()] {
+      return std::make_unique<FilterStateInput<MatchingDataType>>(filter_state_key);
+    };
   };
 
   ProtobufTypes::MessagePtr createEmptyConfigProto() override {
```

Comment:
Seems the HTTP version filter state input reuse the proto of network filter state input. But there are lots of similar cases in the `inputs.h`. So that's fine.

cc @htuch

Reply:
It's a bit weird as you point out. But, I don't think there is a way to selectively pick filter state from network or HTTP stream info when matching, so it seems technically workable.